File: CS 332/p3-Cappers.py
class TreeNode: #Class for creating each node of the tree
    def __init__(self, data):
        self.data = data #This gives the node an integer value
        self.left = None #This gives the node a left child if needed
        self.center = None #This gives the node a center child if needed
        self.right = None #This gives the node a right child if needed

class Tree: #Class for creating the actual tree
    def create_tree_node(data): #This takes a interger value and creates a node with that value for the tree
        return TreeNode(data)

    def fill_tree_node(root): #This function is used to fill the treenodes with the values needed to create the tree
        #creating the left side of the tree
        root.left = Tree.create_tree_node(2)
        root.left.left = Tree.create_tree_node(5)
        root.left.right = Tree.create_tree_node(6)
        root.left.left.left = Tree.create_tree_node(11)
        root.left.left.right = Tree.create_tree_node(12)
        root.left.left.right.left = Tree.create_tree_node(17)

        #creating the center of the tree
        root.center = Tree.create_tree_node(3)
        root.center.left = Tree.create_tree_node(7)
        root.center.right = Tree.create_tree_node(8)

        #creating the right side of the tree
        root.right = Tree.create_tree_node(4)
        root.right.left = Tree.create_tree_node(9)
        root.right.right = Tree.create_tree_node(10)
        root.right.left.left = Tree.create_tree_node(13)
        root.right.left.right = Tree.create_tree_node(14)
        root.right.right.left = Tree.create_tree_node(15)
        root.right.right.right = Tree.create_tree_node(16)
    
    def print_given_level(root, level): #This function is used in the breadth first traversal to find and move up the tree's level
        if root is not None and level == 1: #When the root level of the tree is not empty and the level is 1, it prints just the root value
            print(f"{root.data} ", end="")
        elif root is not None and level > 1: #When the tree has more then just one level it recursively calls itself with the level -1. Then it prints the tree's height 
            Tree.print_given_level(root.left, level - 1)
            Tree.print_given_level(root.center, level - 1) 
            Tree.print_given_level(root.right, level - 1)
        else:
            return 0 #If the tree is empty it returns 0 because the tree is empty
        
    def tree_height(root): #This function is used to find the height of the tree
        if root is None: # If the root is empty then exit since the tree is empty
            return 0
        else: #If the tree is not empty then it recursively calls itself to find the height of the tree
            left_height = Tree.tree_height(root.left)
            center_height = Tree.tree_height(root.center)
            right_height = Tree.tree_height(root.right)

            if left_height > center_height and left_height > right_height: #This is used to find the height of the left side of the tree
                return left_height + 1
            elif center_height > right_height and center_height > left_height: #This is used to find the height of the center of the tree
                return center_height + 1
            else:
                return right_height + 1 #This is used to find the height of the right side of the tree
        
    def bredth_first_traversal(root):
        height = Tree.tree_height(root)
        for i in range(1, height + 1):
            Tree.print_given_level(root, i)

    def depth_first_traversal(root):
        if root is not None:
            Tree.depth_first_traversal(root.left)
            Tree.depth_first_traversal(root.center)
            Tree.depth_first_traversal(root.right)
            print(f"{root.data} ", end="")
            
            
        else:
            return 0


def main(): #This is the main function that is used to run the program. This will create the tree and print the tree's height and the breadth first traversal of the tree
    root = Tree.create_tree_node(1)
    Tree.fill_tree_node(root)
    print("Tree Height: ", Tree.tree_height(root))
    print("\nThe breath first traversal of the tree is: ")
    Tree.bredth_first_traversal(root)
    print("\nThe depth first traversal of the tree is: ")
    Tree.depth_first_traversal(root)

if __name__ == "__main__":
    main()

File: CS 332/p3-Cappers.c
#include <stdio.h>
#include <stdlib.h>

typedef struct treenode {
  int data;
  struct treenode *left;
  struct treenode *right;

} TREE_NODE_TYPE;

TREE_NODE_TYPE *createTreeNode(int data); // function to create the binary tree
void fillTreeNode(TREE_NODE_TYPE *); // function to fill the nodes in the binary tree
void printGivenLevel(TREE_NODE_TYPE *,int); // function to print the given level of the binary tree
void breadthFirstTraversal(TREE_NODE_TYPE *); // function to print the breadth first traversal of the binary tree
void depthFirstTraversal(TREE_NODE_TYPE *); // function to print the depth first traversal of the binary tree
int treeHeight(TREE_NODE_TYPE *); // function to find the height of the binary tree

int main(int argc, char *argv[]) {
  TREE_NODE_TYPE *root = createTreeNode(1);
  fillTreeNode(root);

  printf("The height of the tree is: %d\n", treeHeight(root));
  printf("The breath first traversal of the tree is: ");
  breadthFirstTraversal(root);
  printf("\n");
  printf("The depth first traversal of the tree is: ");
  depthFirstTraversal(root);

  return 0;
}

TREE_NODE_TYPE *createTreeNode(int data) {
  TREE_NODE_TYPE *newTreeNode = (TREE_NODE_TYPE *)malloc(sizeof(TREE_NODE_TYPE));
  newTreeNode -> data = data;
  newTreeNode -> left = NULL;
  newTreeNode -> right = NULL;

  return newTreeNode;
}

void fillTreeNode(TREE_NODE_TYPE *root) {
  // creating the left side of the tree.
  root -> left = createTreeNode(2);
  root -> left -> left = createTreeNode(5);
  root -> left -> right = createTreeNode(6);
  root -> left -> left -> left = createTreeNode(11);
  root -> left -> left -> right = createTreeNode(12);
  root -> left -> left -> right -> left = createTreeNode(17);
/*  // tree should look like this so far:
  //           1
  //         /
  //        2
  //        / \   
  //     5   6
  //      / \
  //   11 12
  //       /
  //      17
  // */

  // creating the right side of the tree.
  root -> right = createTreeNode(4);
  root -> right -> left = createTreeNode(9);
  root -> right -> left -> left = createTreeNode(13);
  root -> right -> left -> right = createTreeNode(14);
  root -> right -> right = createTreeNode(10);
  root -> right -> right -> left = createTreeNode(15);
  root -> right -> right -> right = createTreeNode(16);
  // complete tree should look like this:
  //           1
  //          /   \
  //         2     4
  //        / \   / \
  //       5   6 9   10
  //      / \   / \  / \
  //     11 12 13 14 15 16
  //       / 
  //      17
}

void printGivenLevel(TREE_NODE_TYPE *root, int level){
  if (root == NULL || level < 1){
    return;
  }
  if (level == 1){
    printf("%d, ", root -> data);
  }else{
    printGivenLevel(root -> left, level -1);
    printGivenLevel(root -> right, level -1);
  }
}

void breadthFirstTraversal(TREE_NODE_TYPE *root){
  //values should be outputted from left to right, top to bottom.
  int height = treeHeight(root);
  int i;
  for (i = 1; i <= height; i++){
    printGivenLevel(root, i);
  }
}

void depthFirstTraversal(TREE_NODE_TYPE *root){
  // Needs to out put the values from the bottom up. 
  if (root != NULL){
    depthFirstTraversal(root -> left);
    depthFirstTraversal(root -> right);
    printf("%d, ", root -> data);
  }else{
    return;
  }
}

int treeHeight(TREE_NODE_TYPE *root){
  int heightLeft = 0;
  int heightRight = 0;

  if (root != NULL){
    heightLeft = treeHeight(root -> left);
    heightRight = treeHeight(root -> right);

    return (heightLeft > heightRight) ? (heightLeft + 1) : (heightRight + 1);

  }else {
    return 0;
  }
}